A46-xds-nack-semantics-improvement.md

```diff
@@ -0,0 +1,122 @@
+A46: xDS NACK Semantics Improvement
+----
+* Author(s): Mark D. Roth (markdroth)
+* Approver: ejona86, dfawley
+* Status: {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2021-09-03
+* Discussion at: https://groups.google.com/g/grpc-io/c/gFYDcWIu9B8
+
+## Abstract
+
+This proposal clarifies xDS NACK semantics used in gRPC.
+
+## Background
+
+The [xDS
+spec](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol)
+says that, at the wire protocol level, a NACK is sent on a per-response
+basis, not a per-resource basis.  Although not explicitly stated
+in the spec, this implies that the client must reject (i.e., not
+actually use) even the valid resources from that response, since that
+would allow the server to know the client's actual state (although see
+[Rationale](#rationale) below).  That is the behavior that gRPC currently
+implements.
+
+However, this approach causes problems in a case where a single resource
+is invalid and it causes the client to ignore *all* of the resources in
+the update, especially for LDS and CDS, where the server must send all
+resources in every response.  For example, if there is a single invalid
+Cluster resource in a CDS response, all of the Cluster resources will
+be rejected.  If this happens on the first CDS response after a client
+starts up (i.e., when the client has not already accepted previous
+versions of the CDS resources that it can continue to use), that will
+cause the client to have no valid CDS resources at all, which means
+that the problem will prevent *all* clusters from functioning instead of
+affecting only the invalid resource.  This is particularly problematic
+now that gRPC shares its XdsClient between channels, because a single
+invalid Cluster resource can basically cause all of the client's channels
+to stop working all at once.
+
+This behavior makes it challenging for xDS servers to safely deploy
+changes in environments in which the clients are not centrally controlled.
+For example, older gRPC clients support only the `ROUND_ROBIN`
+LB policy, as per [gRFC A27](A27-xds-global-load-balancing.md),
+but newer clients now support the `RING_HASH` policy, as per [gRFC
+A42](A42-xds-ring-hash-lb-policy.md).  If a Cluster resource specifies
+an unsupported LB policy, clients will consider the resource invalid,
+which will cause them to NACK the response.  So if an xDS server cannot
+be sure that all of its clients have been upgraded to a version that
+supports the `RING_HASH` policy, then it cannot safely send a Cluster
+resource configuring that policy, because that change would cause older
+clients to stop functioning.
+
```

Comment:
We still don't actually have a design for version negotiation, so that's not a solution for the immediate problem.  And even once we do have such a design, I suspect that that's not going to be the best way to address this kind of use-case, because it would require the control plane to have a bunch of logic to determine how to configure older clients that don't support the current configuration.

I think having the control plane generate different versions of the config will make sense for cases where we are deprecating one set of fields and replacing them with another set, where the semantic meaning of the underlying config can be expressed equally well via both sets of fields.  But I think it will be too complex to do in cases where the client simply does not support the features that the configuration is intending to configure, because I don't think the control plane can really know what to do in that case.